pkg/cloudprovider/provider/vultr/provider.go

```diff
@@ -155,28 +156,48 @@ func (p *provider) Validate(ctx context.Context, _ *zap.SugaredLogger, spec clus
 
 	client := getClient(ctx, c.APIKey)
 
-	plans, err := client.Region.Availability(ctx, c.Region, "")
-
-	// TODO: Validate region separately
-	if err != nil {
-		return fmt.Errorf("invalid/not supported region specified %q: %w", c.Region, err)
-	}
-
-	planFound := false
+	if c.MachineType == string(vultrtypes.CloudInstance) || c.MachineType == "" {
+		plans, err := client.Region.Availability(ctx, c.Region, "all")
+		if err != nil {
+			return fmt.Errorf("invalid/not supported region specified %q: %w", c.Region, err)
+		}
+		var planFound bool
+		for _, p := range plans.AvailablePlans {
+			if p == c.Plan {
+				planFound = true
+				break
+			}
+		}
+		if !planFound {
+			return fmt.Errorf("invalid/not supported plan specified %q: %w", c.Plan, err)
+		}
+	} else if c.MachineType == string(vultrtypes.BareMetal) {
+		planAvailability, _, err := client.Plan.ListBareMetal(ctx, nil)
+		if err != nil {
+			return err
+		}
 
-	// Check if given plan present in the returned list
-	for _, plan := range plans.AvailablePlans {
-		if plan == c.Plan {
-			planFound = true
+		var foundPlanRegion bool
+		for _, p := range planAvailability {
+			if p.ID == c.Plan {
+				for _, r := range p.Locations {
+					if r == c.Region {
+						foundPlanRegion = true
+					}
+				}
+			}
 		}
+		if !foundPlanRegion {
+			return fmt.Errorf("plan %q not available on region %q", c.Plan, c.Region)
+		}
+	} else {
+		return fmt.Errorf("unknown machine type %q", c.MachineType)
 	}
-	if !planFound {
-		return fmt.Errorf("invalid/not supported plan specified %q: %w", c.Plan, err)
-	}
+
 	return nil
```

Comment:
I think the latest `return nil` is unreachable and you can just return error from the `else` directly.

Reply:
It actually turns out to be a catch all if none of the branches found an error, so it's reachable for the "everything is fine everywhere" case.